intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,518 @@
+# Basics: named dimensions
+# ------------------------
+#
+# PyTorch now allows Tensors to have named dimensions; factory functions
+# now take a new `names` argument that associates a name with each dimension.
+# This works with most factory functions, such as
+#
+# - `tensor`
+# - `empty`
+# - `ones`
+# - `zeros`
+# - `randn`
+# - `rand`
+#
+# Here we construct a tensor with names:
+
+import torch
+imgs = torch.randn(1, 2, 2, 3 , names=('N', 'C', 'H', 'W'))
+
+# Unlike in
+# `the original named tensors blogpost <http://nlp.seas.harvard.edu/NamedTensor>`_,
+# named dimensions are ordered: `tensor.names[i]` is the name of the `i`th dimension of `tensor`.
+#
+# There are two ways rename a ``Tensor``'s dimensions:
+#
+
+print(imgs.names)
+
+# Method #1: set .names attribute
+imgs.names = ['batch', 'channel', 'width', 'height']
+print(imgs.names)
+
+# Method #2: specify new names:
+imgs.rename(channel='C', width='W', height='H')
+print(imgs.names)
+
+######################################################################
+# The preferred way to remove names is to call ``tensor.rename(None)``:
+
+imgs.rename(None)
+print(imgs.names)
+
+######################################################################
+# Unnamed tensors (tensors with no named dimensions) still work as normal and do
+# not have names in their `repr`.
+
+unnamed = torch.randn(2, 1, 3)
+print(unnamed)
+print(unnamed.names)
+
+######################################################################
+# Named tensors do not require that all dimensions be named.
+
+imgs = torch.randn(3, 1, 1, 2, names=('N', None, None, None))
+print(imgs.names)
+
+######################################################################
+# Because named tensors coexist with unnamed tensors, we need a nice way to write named-tensor-aware
+# code that works with both named and unnamed tensors. Use ``tensor.refine_names(*names)`` to refine
+# dimensions and lift unnamed dims to named dims. Refining a dimension is defined as a "rename" with
+# the following constraints:
+#
+# - A ``None`` dim can be refined to have any name
+# - A named dim can only be refined to have the same name.
+
+imgs = torch.randn(3, 1, 1, 2)
+named_imgs= imgs.refine_names('N', 'C', 'H', 'W')
+print(named_imgs.names)
+
+# Coerces the last two dims to 'H' and 'W'. In Python 2, use the string '...' instead of ...
+print(imgs.refine_names(..., 'H', 'W').names)
+
+def catch_error(fn):
+    try:
+        fn()
+    except RuntimeError as err:
+        print(err)
+
+# Tried to refine an existing name to a different name
+print(catch_error(lambda: imgs.refine_names('batch', 'channel', 'height', 'width')))
+
+######################################################################
+# Most simple operations propagate names. The ultimate goal for named tensors is
+# for all operations to propagate names in a reasonable, intuitive manner. Many
+# common operations have been added at the time of the 1.3 release; here,

```

Comment:
Would be good to link to the docs, but this is a chicken and egg problem...

Reply:
We'll keep as a to-do to do this after the docs land.